### PR TITLE
infra docker files

### PR DESCRIPTION
**Pull request by kobymeir**
## Status
Done

## Description
moving infra files

fixes: https://jira-dc.paloaltonetworks.com/browse/CIAC-7077

### CODE REVIEW

File: .gitlab/helper_functions.sh [hunk 1]
```diff
@@ -0,0 +1,81 @@
+#!/bin/bash
+
+CYAN="\e[0;36m"
+CLEAR="\e[0m"
+SECTION_START="\e[0Ksection_start:the_time:section_id\r\e[0K${CYAN}section_header${CLEAR}"
+SECTION_END="\e[0Ksection_end:the_time:section_id\r\e[0K"
+RED='\033[0;31m'
+GREEN='\033[0;32m'
+BLUE='\033[0;34m'
+NC='\033[0m'
```
Comment (tkatzir) [inline .gitlab/helper_functions.sh:10]:
```suggestion
RED='\033[0;31m'
GREEN='\033[0;32m'
BLUE='\033[0;34m'
NC='\033[0m'

SECTION_START="\e[0Ksection_start:the_time:section_id\r\e[0K${CYAN}section_header${CLEAR}"
SECTION_END="\e[0Ksection_end:the_time:section_id\r\e[0K"
```

File: .gitlab/helper_functions.sh [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+section_start() {
+    local section_header section_id start
+    start="$SECTION_START"
+    if [[ "$#" -eq 1 ]]; then
+        section_header="$1"
+        section_id="$(echo "$1" | tr -c '[:alnum:]\n\r' '_')"
```
Comment (tkatzir) [inline .gitlab/helper_functions.sh:17]: `\n\r` or `\r\n` ? (i.e. `CRLF` or `LFCR`?)
In several places.

File: .gitlab/helper_functions.sh [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        else
+            section_header="$2"
+            section_id="$1"
+        fi
+    elif [[ "$#" -eq 3 && "$3" =~ /^-{0,2}collapsed$/ ]]; then
+        start="${start/section_id/section_id[collapsed=true]}"
+        section_header="$2"
+        section_id="$1"
+    else
+        echo "section_start should be called with 1-3 args but it was called with $#"
```
Comment (tkatzir) [inline .gitlab/helper_functions.sh:32]:
```suggestion
        echo "section_start should be called with 1-3 args but it was called with $# args"
```

File: .gitlab/helper_functions.sh [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        echo "acceptable usages:"
+        echo "    1. section_start \"<section-start-id>\" \"<section-header>\""
+        echo "    2. section_start \"<section-header>\""
+        echo "    3. section_start \"<section-start-id>\" \"<section-header>\" --collapse"
+        echo "    4. section_start \"<section-header>\" --collapse"
```
Comment (tkatzir) [inline .gitlab/helper_functions.sh:37]: Suggesting. Why not in order?
```suggestion
        echo "    1. section_start \"<section-header>\""
        echo "    2. section_start \"<section-start-id>\" \"<section-header>\""
        echo "    3. section_start \"<section-header>\" --collapse"
        echo "    4. section_start \"<section-start-id>\" \"<section-header>\" --collapse"
```

File: .gitlab/helper_functions.sh [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+section_end() {
+    local section_id end
+    date +"[%Y-%m-%dT%H:%M:%S.%3N] section end"
+    end="$SECTION_END"
+    if [[ "$#" -eq 1 ]]; then
+        section_id="$(echo "$1" | tr -c '[:alnum:]\n\r' '_')"
+    else
+        echo "section_end should be called with 1 arg but it was called with $#"
```
Comment (tkatzir) [inline .gitlab/helper_functions.sh:55]:
```suggestion
        echo "section_end should be called with 1 arg but it was called with $# args"
```

File: .gitlab/helper_functions.sh [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        echo "acceptable usage:"
+        echo "    1. section_end \"<section-start-id>\""
+        echo "    2. section_start \"<section-header>\""
```
Comment (tkatzir) [inline .gitlab/helper_functions.sh:58]: Copy-Paste?
```suggestion
        echo "    2. section_end \"<section-header>\""
```

File: .gitlab/helper_functions.sh [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        echo "where <section-start-id> or <section-header> is that of the section this marks the end of"
```
Comment (tkatzir) [inline .gitlab/helper_functions.sh:59]:
```suggestion
        echo "where <section-start-id> or <section-header> is the id of the section this marks the end of"
```

File: .hooks/bootstrap [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    cat >&2 <<__EOF__
+ERROR: this script must be run at the root of the source tree
+__EOF__
+    exit 1
+fi
+
+echo "======================="
+if [ -z "${INSTALL_POETRY}" ]; then
+    if ! command -v poetry >/dev/null 2>&1; then
+        echo "ERROR: poetry is missing. Please run the following command to install poetry:
```
Comment (tkatzir) [inline .hooks/bootstrap:38]:
```suggestion
        echo "ERROR: poetry is missing. Please run the following command to install it:
```

File: .hooks/bootstrap [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+               curl -sSL https://install.python-poetry.org | python3 -" 1>&2
+        exit 1
+    fi
+else
+    should_install_poetry="yes"
+    if command -v poetry >/dev/null 2>&1; then
+        if [[ "$(poetry --version)" == "Poetry (version ${POETRY_VERSION})" ]]; then
+            echo "Poetry already installed:$(poetry --version) with correct version"
```
Comment (tkatzir) [inline .hooks/bootstrap:46]: Maybe?
```suggestion
            echo "Poetry is already installed with the correct version: $(poetry --version)"
```

File: .hooks/bootstrap [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+            should_install_poetry="no"
+        else
+            echo "Poetry installed with a different version:$(poetry --version) required:${POETRY_VERSION}"
```
Comment (tkatzir) [inline .hooks/bootstrap:49]: Maybe better to align with the message in line 64?
```suggestion
            echo "Poetry is already installed with a different version: $(poetry --version), required version: ${POETRY_VERSION}"
```

File: .hooks/bootstrap [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        exit_on_error $error_code "Failed to install Poetry version:${POETRY_VERSION}"
+    fi
+fi
+
+GIT_HOOKS_DIR="${PWD}/.git/hooks"
+
+if [ -n "${NO_HOOKS}" ]; then
+    echo "Skipping hooks setup as environment variable NO_HOOKS is set"
+else
+    if [ ! -e "${GIT_HOOKS_DIR}/pre-commit" ]; then
```
Comment (tkatzir) [inline .hooks/bootstrap:78]: Define `GIT_HOOKS_DIR ` only if it's needed
```suggestion
if [ -n "${NO_HOOKS}" ]; then
    echo "Skipping hooks setup as environment variable NO_HOOKS is set"
else
    GIT_HOOKS_DIR="${PWD}/.git/hooks"
    if [ ! -e "${GIT_HOOKS_DIR}/pre-commit" ]; then
```

File: .hooks/bootstrap [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    exit_on_error $? "Failed to check poetry files"
+    echo "Installing dependencies..."
+    poetry install
+    exit_on_error $? "Failed to install dependencies"
+fi
+
+echo "=========================="
+echo "Done setting up virtualenv with poetry"
+echo "Activate the venv by running: poetry shell"
+echo "Deactivate by running: deactivate"
```
Comment (tkatzir) [inline .hooks/bootstrap:110]: Standardize. either `virtualenv` or `venv`
```suggestion
echo "Done setting up virtualenv with poetry"
echo "Activate the virtualenv by running: poetry shell"
echo "Deactivate by running: deactivate"
```

File: .pre-commit-config.yaml [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+- repo: https://github.com/pre-commit/pre-commit-hooks
+  rev: v4.6.0
+  hooks:
+  - id: check-json
+  - id: check-yaml
+    exclude: .gitlab
+  - id: check-ast
+  - id: check-merge-conflict
+#  - id: debug-statements
+#    language_version: python3
```
Comment (tkatzir) [inline .pre-commit-config.yaml:23]: Needed?

File: docker/build_docker.sh [hunk 14]
```diff
@@ -356,44 +367,34 @@ function docker_build {
...
-
-if [[ ! $(which pyenv) ]] && [[ -n "${CIRCLECI}" ]]; then 
-    echo "pyenv not found. setting up necessary env for pyenv on circle ci";\
+if [[ ! $(which pyenv) ]] && [[ -n "${GITLAB_CI}" ]]; then
+    echo "pyenv not found. setting up necessary env for pyenv on CI";\
     export PATH="$HOME/.pyenv/bin:$PATH"
     eval "$(pyenv init -)"
     eval "$(pyenv virtualenv-init -)"
-    pyenv shell system $(pyenv versions --bare | grep 3.10)
+    pyenv shell system "$(pyenv versions --bare | grep 3.10)"
```
Comment (tkatzir) [inline docker/build_docker.sh:397]: Can the magic value of `3.10` be set at the top of the file, or somewhere where it stands out?

File: docker/python3/Dockerfile [hunk 15]
```diff
@@ -1,6 +1,8 @@
 # Last modified: 2024-11-14T00:16:47.268736+00:00
 FROM python:3.11.10-alpine3.20
 
+
+
```
Comment (tkatzir) [inline docker/python3/Dockerfile:5]: Why 2 extra lines?

**Review by tkatzir — APPROVED** [review]
Great!
Approved, with minor comments.

File: .gitlab/helper_functions.sh (hunk 2 above)
Comment (kobymeir) [reply .gitlab/helper_functions.sh:17]: the order is not important here in `tr`

File: .pre-commit-config.yaml (hunk 13 above)
Comment (kobymeir) [inline .pre-commit-config.yaml:23]:
```suggestion
```

File: docker/build_docker.sh [hunk 16]
```diff
@@ -3,25 +3,29 @@
 # exit on errors
 set -e
 
-REVISION=${CIRCLE_BUILD_NUM:-`date +%s`}
-PUSHED_DOCKERS=''
-CURRENT_DIR=`pwd`
-SCRIPT_DIR=$(dirname ${BASH_SOURCE})
+REVISION=${CI_PIPELINE_ID:-$(date +%s)}
```
[the remark this answers is not in the document]
Comment (kobymeir) [reply docker/build_docker.sh:6]: the number is higher than circle, so there should be no conflict.

File: docker/python3/Dockerfile (hunk 15 above)
Comment (kobymeir) [reply docker/python3/Dockerfile:5]: that's how i do my testing, by changing the image i'm causing it to be built, will be removed befre the merge

File: pyproject.toml [hunk 17]
```diff
@@ -0,0 +1,41 @@
+[build-system]
+requires = ["poetry-core"]
```
[the remark this answers is not in the document]
Comment (kobymeir) [reply pyproject.toml:2]: it will replace the pipenv file. i will remove the pipenv after we're done with this PR.

**Pull request merged**